File: Práctica2/laberinto.py
import os
import msvcrt
import collections
import heapq
import copy

# --- Constantes ---
N = 20

# --- Tipos de Casilla (usando constantes) ---
VACIO = 0
MURO = 1
INICIO = 2
FIN = 3
CAMINO_BFS = 4
CAMINO_DFS = 5
CAMINO_ASTAR = 6

# --- Colores ANSI ---
class Colors:
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    MAGENTA = '\033[95m'
    RESET = '\033[0m'

class Laberinto:
    def __init__(self, n):
        self.n = n
        self.matriz = [[VACIO for _ in range(n)] for _ in range(n)]
        self.inicio = None
        self.fin = None
        self.cursor_x = 0
        self.cursor_y = 0

    def dibujar(self, titulo):
        os.system('cls' if os.name == 'nt' else 'clear')
        print(titulo)
        for y in range(self.n):
            linea = ""
            for x in range(self.n):
                if x == self.cursor_x and y == self.cursor_y:
                    linea += "[]"
                else:
                    casilla = self.matriz[y][x]
                    if casilla == MURO:
                        linea += "##"
                    elif casilla == INICIO:
                        linea += "IN"
                    elif casilla == FIN:
                        linea += "FN"
                    elif casilla == CAMINO_BFS:
                        linea += f"{Colors.GREEN}**{Colors.RESET}"
                    elif casilla == CAMINO_DFS:
                        linea += f"{Colors.YELLOW}++{Colors.RESET}"
                    elif casilla == CAMINO_ASTAR:
                        linea += f"{Colors.MAGENTA}@@{Colors.RESET}"
                    else:
                        linea += " ."
            print(linea)
        
        if self.cursor_x is not None:
            print("\n--- MODO EDICION ---")
            print("Usa W/A/S/D para moverte.")
            print("ESPACIO: Muro | I: Inicio | F: Fin")
            print("Q: Guardar y Resolver con todos los algoritmos.")

    def modo_edicion(self):
        self.dibujar("Construye tu laberinto")
        while True:
            tecla = msvcrt.getch().lower()
            if tecla == b'w' and self.cursor_y > 0: self.cursor_y -= 1
            elif tecla == b's' and self.cursor_y < self.n - 1: self.cursor_y += 1
            elif tecla == b'a' and self.cursor_x > 0: self.cursor_x -= 1
            elif tecla == b'd' and self.cursor_x < self.n - 1: self.cursor_x += 1
            elif tecla == b' ':
                if self.matriz[self.cursor_y][self.cursor_x] == VACIO:
                    self.matriz[self.cursor_y][self.cursor_x] = MURO
                elif self.matriz[self.cursor_y][self.cursor_x] == MURO:
                    self.matriz[self.cursor_y][self.cursor_x] = VACIO
            elif tecla == b'i':
                if self.inicio:
                    self.matriz[self.inicio[1]][self.inicio[0]] = VACIO
                self.matriz[self.cursor_y][self.cursor_x] = INICIO
                self.inicio = (self.cursor_x, self.cursor_y)
            elif tecla == b'f':
                if self.fin:
                    self.matriz[self.fin[1]][self.fin[0]] = VACIO
                self.matriz[self.cursor_y][self.cursor_x] = FIN
                self.fin = (self.cursor_x, self.cursor_y)
            elif tecla == b'q':
                if not self.inicio or not self.fin:
                    print("\nError: Debes definir un INICIO ('I') y un FIN ('F') para resolver.")
                    return False
                self.cursor_x = None # Salir del modo edición
                self.cursor_y = None
                return True
            self.dibujar("Construye tu laberinto")

    def resolver_bfs(self):
        matriz_copia = copy.deepcopy(self.matriz)
        frontera = collections.deque([self.inicio])
        visitados = {self.inicio}
        padres = {self.inicio: None}
        solucion_encontrada = False

        while frontera:
            x, y = frontera.popleft()
            if (x, y) == self.fin:
                solucion_encontrada = True
                break
            
            for dx, dy in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                nx, ny = x + dx, y + dy
                if 0 <= nx < self.n and 0 <= ny < self.n and (nx, ny) not in visitados and matriz_copia[ny][nx] != MURO:
                    visitados.add((nx, ny))
                    padres[(nx, ny)] = (x, y)
                    frontera.append((nx, ny))

        if solucion_encontrada:
            actual = self.fin
            while actual != self.inicio:
                x, y = actual
                if matriz_copia[y][x] not in [INICIO, FIN]:
                    matriz_copia[y][x] = CAMINO_BFS
                actual = padres[actual]
        
        self.matriz = matriz_copia
        self.dibujar("--- Solucion con BFS (Camino mas corto: **) ---" if solucion_encontrada else "--- BFS: No se encontro solucion ---")

    def resolver_dfs(self):
        matriz_copia = copy.deepcopy(self.matriz)
        visitados = set()
        
        def dfs_recursivo(x, y):
            if not (0 <= x < self.n and 0 <= y < self.n) or (x, y) in visitados or matriz_copia[y][x] == MURO:
                return False
            
            visitados.add((x, y))
            
            if (x, y) == self.fin:
                return True

            for dx, dy in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                if dfs_recursivo(x + dx, y + dy):
                    if matriz_copia[y][x] != INICIO:
                        matriz_copia[y][x] = CAMINO_DFS
                    return True
            return False

        solucion_encontrada = dfs_recursivo(self.inicio[0], self.inicio[1])
        self.matriz = matriz_copia
        self.dibujar("--- Solucion con DFS (Primer camino encontrado: ++) ---" if solucion_encontrada else "--- DFS: No se encontro solucion ---")

    def resolver_a_star(self):
        matriz_copia = copy.deepcopy(self.matriz)
        
        def manhattan(a, b):
            return abs(a[0] - b[0]) + abs(a[1] - b[1])

        frontera = [(0, self.inicio)] # (f_score, (x, y))
        padres = {self.inicio: None}
        g_score = { (x,y): float('inf') for y in range(self.n) for x in range(self.n) }
        g_score[self.inicio] = 0
        f_score = { (x,y): float('inf') for y in range(self.n) for x in range(self.n) }
        f_score[self.inicio] = manhattan(self.inicio, self.fin)
        
        solucion_encontrada = False

        while frontera:
            _, actual = heapq.heappop(frontera)
            x, y = actual

            if actual == self.fin:
                solucion_encontrada = True
                break

            for dx, dy in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                vecino = (x + dx, y + dy)
                nx, ny = vecino

                if not (0 <= nx < self.n and 0 <= ny < self.n) or matriz_copia[ny][nx] == MURO:
                    continue
                
                temp_g = g_score[actual] + 1
                if temp_g < g_score[vecino]:
                    padres[vecino] = actual
                    g_score[vecino] = temp_g
                    f_score[vecino] = temp_g + manhattan(vecino, self.fin)
                    heapq.heappush(frontera, (f_score[vecino], vecino))
        
        if solucion_encontrada:
            actual = self.fin
            while actual != self.inicio:
                x, y = actual
                if matriz_copia[y][x] not in [INICIO, FIN]:
                    matriz_copia[y][x] = CAMINO_ASTAR
                actual = padres[actual]

        self.matriz = matriz_copia
        self.dibujar("--- Solucion con A* (Camino Optimo: @@) ---" if solucion_encontrada else "--- A*: No se encontro solucion ---")

if __name__ == "__main__":
    laberinto = Laberinto(N)
    
    if laberinto.modo_edicion():
        laberinto.resolver_bfs()
        input("\nPresiona Enter para ver la solucion con DFS...")
        
        laberinto_original = copy.deepcopy(laberinto.matriz) # Guardar estado original
        laberinto.resolver_dfs()
        input("\nPresiona Enter para ver la solucion con A*...")

        laberinto.matriz = laberinto_original # Restaurar para A*
        laberinto.resolver_a_star()

        print(f"{Colors.RESET}\nPrograma terminado.")

File: Práctica2/laberinto.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <conio.h>
#include <windows.h>
#include <limits.h> // Para INT_MAX

#define N 20

// --- Definiciones para los tipos de casilla ---
#define VACIO 0
#define MURO 1
#define INICIO 2
#define FIN 3
#define CAMINO_BFS 4 // Verde
#define CAMINO_DFS 5 // Amarillo
#define CAMINO_ASTAR 6 // Magenta

// --- Estructuras para Nodo y Cola (usadas por BFS) ---
typedef struct {
    int x, y;
} Nodo;

typedef struct {
    Nodo items[N * N];
    int frente, final;
} Cola;

// --- Prototipos de funciones ---
void setColor(int color);
void dibujarMatriz(int matriz[N][N], int cX, int cY, const char* titulo);

// --- Función para cambiar el color de la consola ---
void setColor(int color) {
    HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
    SetConsoleTextAttribute(hConsole, color);
}

// --- Funciones de la Cola para BFS ---
void inicializarCola(Cola *q) { q->frente = -1; q->final = -1; }
int estaVacia(Cola *q) { return q->frente == -1; }
void encolar(Cola *q, Nodo valor) {
    if (q->final == N * N - 1) return;
    if (q->frente == -1) q->frente = 0;
    q->final++;
    q->items[q->final] = valor;
}
Nodo desencolar(Cola *q) {
    Nodo item = {-1, -1};
    if (estaVacia(q)) return item;
    item = q->items[q->frente];
    q->frente++;
    if (q->frente > q->final) q->frente = q->final = -1;
    return item;
}

// --- Función de Dibujo (actualizada para todos los caminos) ---
void dibujarMatriz(int matriz[N][N], int cursorX, int cursorY, const char* titulo) {
    system("cls");
    printf("%s\n", titulo);
    for (int y = 0; y < N; y++) {
        for (int x = 0; x < N; x++) {
            if (x == cursorX && y == cursorY) printf("[]");
            else {
                switch (matriz[y][x]) {
                    case MURO:       printf("##"); break;
                    case INICIO:     printf("IN"); break;
                    case FIN:        printf("FN"); break;
                    case CAMINO_BFS: setColor(10); printf("**"); setColor(7); break;
                    case CAMINO_DFS: setColor(14); printf("++"); setColor(7); break;
                    case CAMINO_ASTAR:setColor(13); printf("@@"); setColor(7); break;
                    default:         printf(" ."); break;
                }
            }
        }
        printf("\n");
    }
    if (cursorX != -1) {
        printf("\n--- MODO EDICION ---");
        printf("\nUsa W/A/S/D para moverte.");
        printf("\nESPACIO: Muro | I: Inicio | F: Fin");
        printf("\nQ: Guardar y Resolver con todos los algoritmos.\n");
    }
}

// --- ALGORITMOS DE BÚSQUEDA ---

// BFS (sin cambios)
void resolverLaberintoBFS(int matriz[N][N], Nodo inicio, Nodo fin);

// DFS (sin cambios)
void resolverLaberintoDFS(int matriz[N][N], Nodo inicio, Nodo fin);
int resolverDFS_recursivo(int matriz[N][N], int v[N][N], Nodo a, Nodo f);

// --- A* (A-Estrella) ---
int calcularManhattan(Nodo a, Nodo b) {
    return abs(a.x - b.x) + abs(a.y - b.y);
}

void resolverLaberintoAStar(int matriz[N][N], Nodo inicio, Nodo fin) {
    Nodo frontera[N * N];
    int frontera_count = 0;
    
    int en_frontera[N][N] = {0};
    int visitados[N][N] = {0};

    Nodo padres[N][N];
    memset(padres, -1, sizeof(padres));

    int g_score[N][N];
    int f_score[N][N];
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            g_score[i][j] = INT_MAX;
            f_score[i][j] = INT_MAX;
        }
    }

    g_score[inicio.y][inicio.x] = 0;
    f_score[inicio.y][inicio.x] = calcularManhattan(inicio, fin);

    frontera[frontera_count++] = inicio;
    en_frontera[inicio.y][inicio.x] = 1;

    int solucionEncontrada = 0;

    while (frontera_count > 0) {
        int actual_idx = 0;
        for (int i = 1; i < frontera_count; i++) {
            if (f_score[frontera[i].y][frontera[i].x] < f_score[frontera[actual_idx].y][frontera[actual_idx].x]) {
                actual_idx = i;
            }
        }
        Nodo actual = frontera[actual_idx];

        if (actual.x == fin.x && actual.y == fin.y) {
            solucionEncontrada = 1;
            break;
        }

        // Mover de frontera a visitados
        frontera[actual_idx] = frontera[--frontera_count];
        en_frontera[actual.y][actual.x] = 0;
        visitados[actual.y][actual.x] = 1;

        int dx[] = {0, 0, 1, -1};
        int dy[] = {1, -1, 0, 0};

        for (int i = 0; i < 4; i++) {
            Nodo vecino = {actual.x + dx[i], actual.y + dy[i]};
            
            if (vecino.x < 0 || vecino.x >= N || vecino.y < 0 || vecino.y >= N) continue;
            if (matriz[vecino.y][vecino.x] == MURO || visitados[vecino.y][vecino.x]) continue;

            int temp_g_score = g_score[actual.y][actual.x] + 1;

            if (temp_g_score < g_score[vecino.y][vecino.x]) {
                padres[vecino.y][vecino.x] = actual;
                g_score[vecino.y][vecino.x] = temp_g_score;
                f_score[vecino.y][vecino.x] = temp_g_score + calcularManhattan(vecino, fin);
                
                if (!en_frontera[vecino.y][vecino.x]) {
                    frontera[frontera_count++] = vecino;
                    en_frontera[vecino.y][vecino.x] = 1;
                }
            }
        }
    }

    if (solucionEncontrada) {
        Nodo actual = fin;
        while (actual.x != inicio.x || actual.y != inicio.y) {
            if (matriz[actual.y][actual.x] != FIN) matriz[actual.y][actual.x] = CAMINO_ASTAR;
            actual = padres[actual.y][actual.x];
        }
    }
    dibujarMatriz(matriz, -1, -1, solucionEncontrada ? "--- Solucion con A* (Camino Optimo: @@) ---" : "--- A*: No se encontro solucion ---");
}

// --- MAIN ---
int main() {
    int matriz_original[N][N] = {0};
    int cursorX = 0, cursorY = 0;
    Nodo inicio = {-1, -1}, fin = {-1, -1};
    char tecla;

    dibujarMatriz(matriz_original, cursorX, cursorY, "Construye tu laberinto");

    while ((tecla = _getch()) != 'q' && tecla != 'Q') {
        switch (tecla) {
            case 'w': case 'W': if (cursorY > 0) cursorY--; break;
            case 's': case 'S': if (cursorY < N - 1) cursorY++; break;
            case 'a': case 'A': if (cursorX > 0) cursorX--; break;
            case 'd': case 'D': if (cursorX < N - 1) cursorX++; break;
            case ' ':
                if (matriz_original[cursorY][cursorX] == VACIO) matriz_original[cursorY][cursorX] = MURO;
                else if (matriz_original[cursorY][cursorX] == MURO) matriz_original[cursorY][cursorX] = VACIO;
                break;
            case 'i': case 'I':
                if (inicio.x != -1) matriz_original[inicio.y][inicio.x] = VACIO;
                matriz_original[cursorY][cursorX] = INICIO;
                inicio = (Nodo){cursorX, cursorY};
                break;
            case 'f': case 'F':
                if (fin.x != -1) matriz_original[fin.y][fin.x] = VACIO;
                matriz_original[cursorY][cursorX] = FIN;
                fin = (Nodo){cursorX, cursorY};
                break;
        }
        dibujarMatriz(matriz_original, cursorX, cursorY, "Construye tu laberinto");
    }

    if (inicio.x == -1 || fin.x == -1) {
        printf("\nError: Debes definir un INICIO ('I') y un FIN ('F') para resolver.\n");
        return 1;
    }

    // --- Resolver con BFS ---
    int matriz_bfs[N][N];
    memcpy(matriz_bfs, matriz_original, sizeof(matriz_original));
    resolverLaberintoBFS(matriz_bfs, inicio, fin);
    printf("\nPresiona una tecla para ver la solucion con DFS...");
    _getch();

    // --- Resolver con DFS ---
    int matriz_dfs[N][N];
    memcpy(matriz_dfs, matriz_original, sizeof(matriz_original));
    resolverLaberintoDFS(matriz_dfs, inicio, fin);
    printf("\nPresiona una tecla para ver la solucion con A*...");
    _getch();

    // --- Resolver con A* ---
    int matriz_astar[N][N];
    memcpy(matriz_astar, matriz_original, sizeof(matriz_original));
    resolverLaberintoAStar(matriz_astar, inicio, fin);

    setColor(7);
    printf("\nPrograma terminado.\n");
    return 0;
}


// --- Implementaciones de BFS y DFS (sin cambios, necesarias para compilar) ---

void resolverLaberintoBFS(int matriz[N][N], Nodo inicio, Nodo fin) {
    Cola nodos_frontera;
    inicializarCola(&nodos_frontera);
    int visitados[N][N]; memset(visitados, 0, sizeof(visitados));
    Nodo padres[N][N]; memset(padres, -1, sizeof(padres));
    encolar(&nodos_frontera, inicio);
    visitados[inicio.y][inicio.x] = 1;
    int solucionEncontrada = 0;
    while (!estaVacia(&nodos_frontera)) {
        Nodo nodo_actual = desencolar(&nodos_frontera);
        if (nodo_actual.x == fin.x && nodo_actual.y == fin.y) { solucionEncontrada = 1; break; }
        int dx[] = {0, 0, 1, -1}, dy[] = {1, -1, 0, 0};
        for (int i = 0; i < 4; i++) {
            Nodo hijo = {nodo_actual.x + dx[i], nodo_actual.y + dy[i]};
            if (hijo.x >= 0 && hijo.x < N && hijo.y >= 0 && hijo.y < N) {
                if (!visitados[hijo.y][hijo.x] && matriz[hijo.y][hijo.x] != MURO) {
                    visitados[hijo.y][hijo.x] = 1;
                    padres[hijo.y][hijo.x] = nodo_actual;
                    encolar(&nodos_frontera, hijo);
                }
            }
        }
    }
    if (solucionEncontrada) {
        Nodo actual = fin;
        while (actual.x != inicio.x || actual.y != inicio.y) {
            if(matriz[actual.y][actual.x] != FIN) matriz[actual.y][actual.x] = CAMINO_BFS;
            actual = padres[actual.y][actual.x];
        }
    }
    dibujarMatriz(matriz, -1, -1, solucionEncontrada ? "--- Solucion con BFS (Camino mas corto: **) ---" : "--- BFS: No se encontro solucion ---");
}

int resolverDFS_recursivo(int matriz[N][N], int visitados[N][N], Nodo actual, Nodo fin) {
    if (actual.x<0 || actual.x>=N || actual.y<0 || actual.y>=N || matriz[actual.y][actual.x]==MURO || visitados[actual.y][actual.x]) return 0;
    visitados[actual.y][actual.x] = 1;
    if (actual.x == fin.x && actual.y == fin.y) return 1;
    int dx[] = {0, 0, 1, -1}, dy[] = {1, -1, 0, 0};
    for (int i = 0; i < 4; i++) {
        Nodo siguiente = {actual.x + dx[i], actual.y + dy[i]};
        if (resolverDFS_recursivo(matriz, visitados, siguiente, fin)) {
            if (matriz[actual.y][actual.x] != INICIO) matriz[actual.y][actual.x] = CAMINO_DFS;
            return 1;
        }
    }
    return 0;
}

void resolverLaberintoDFS(int matriz[N][N], Nodo inicio, Nodo fin) {
    int visitados[N][N];
    memset(visitados, 0, sizeof(visitados));
    int solucionEncontrada = resolverDFS_recursivo(matriz, visitados, inicio, fin);
    dibujarMatriz(matriz, -1, -1, solucionEncontrada ? "--- Solucion con DFS (Primer camino encontrado: ++) ---" : "--- DFS: No se encontro solucion ---");
}